internal/controllers/insights_controller.go

```diff
@@ -0,0 +1,129 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controllers
+
+import (
+	"context"
+	"errors"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	"github.com/RedHatInsights/runtimes-inventory-operator/internal/common"
+	"github.com/go-logr/logr"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// InsightsReconciler reconciles the Insights proxy for Cryostat agents
+type InsightsReconciler struct {
+	*InsightsReconcilerConfig
+	backendDomain string
+	proxyDomain   string
+	proxyImageTag string
+}
+
+// InsightsReconcilerConfig contains configuration to create an InsightsReconciler
+type InsightsReconcilerConfig struct {
+	client.Client
+	Log             logr.Logger
+	Scheme          *runtime.Scheme
+	Namespace       string
+	UserAgentPrefix string
+	common.OSUtils
+}
+
+// NewInsightsReconciler creates an InsightsReconciler using the provided configuration
+func NewInsightsReconciler(config *InsightsReconcilerConfig) (*InsightsReconciler, error) {
+	backendDomain := config.GetEnv(common.EnvInsightsBackendDomain)
+	if len(backendDomain) == 0 {
+		return nil, errors.New("no backend domain provided for Insights")
+	}
+	imageTag := config.GetEnv(common.EnvInsightsProxyImageTag)
+	if len(imageTag) == 0 {
+		return nil, errors.New("no proxy image tag provided for Insights")
+	}
+	proxyDomain := config.GetEnv(common.EnvInsightsProxyDomain)
+
+	return &InsightsReconciler{
+		InsightsReconcilerConfig: config,
+		backendDomain:            backendDomain,
+		proxyDomain:              proxyDomain,
+		proxyImageTag:            imageTag,
+	}, nil
+}
+
+// +kubebuilder:rbac:groups=apps,resources=deployments;deployments/finalizers,verbs=*
+// +kubebuilder:rbac:groups="",resources=services;secrets;configmaps;configmaps/finalizers,verbs=*
+// +kubebuilder:rbac:groups=config.openshift.io,resources=clusterversions,verbs=get;list;watch
+
+// Reconcile processes the Insights proxy deployment and configures it accordingly
+func (r *InsightsReconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	reqLogger := r.Log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling Insights Proxy")
+
+	// Reconcile all Insights support
+	err := r.reconcileInsights(ctx)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	return reconcile.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *InsightsReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	c := ctrl.NewControllerManagedBy(mgr).
+		Named("insights").
+		// Filter controller to watch only specific objects we care about
+		Watches(&corev1.Secret{},
```

Comment:
Shouldn't be this just owned? (there are anyways owner references) - it is also good this way, just asking :)
I see that this also works as a filtering

Reply:
Just on sidenote: there are explicit filters also. But yes this is fine also this way.

Reply:
I think there is a fair amount of flexibility with cache filtering, but this would apply to the whole manager:
https://github.com/kubernetes-sigs/controller-runtime/blob/main/designs/cache_options.md

We might be able to short-circuit the watches a bit earlier by using predicate functions, they are called just before handing off to the handler function: https://github.com/kubernetes-sigs/controller-runtime/blob/1382d7d6774050cd844b7f0c3cd63d5258c1b714/pkg/internal/source/event_handler.go#L79-L88. I'm not sure how much of a difference this will make in terms of resource utilization.

Something interesting is using `WatchesMetadata` instead? I think this should reduce the amount of data stored in the cache. What do you think?
https://github.com/kubernetes-sigs/controller-runtime/blob/1382d7d6774050cd844b7f0c3cd63d5258c1b714/pkg/builder/controller.go#L142-L172

Reply:
> Something interesting is using WatchesMetadata instead? I think this should reduce the amount of data stored in the cache. What do you think?

But for the main case the whole resource is needed. Not sure if this would work.

> I think there is a fair amount of flexibility with cache filtering, but this would apply to the whole manager:
https://github.com/kubernetes-sigs/controller-runtime/blob/main/designs/cache_options.md

Wow this is interesting why is that only on this level. Will also take a look.


> We might be able to short-circuit the watches a bit earlier by using predicate functions, they are called just before handing off to the handler function: https://github.com/kubernetes-sigs/controller-runtime/blob/1382d7d6774050cd844b7f0c3cd63d5258c1b714/pkg/internal/source/event_handler.go#L79-L88. I'm not sure how much of a difference this will make in terms of resource utilization.

Yes, that is useful, but resources still will be in the cache.

Reply:
> > Something interesting is using WatchesMetadata instead? I think this should reduce the amount of data stored in the cache. What do you think?
> 
> But for the main case the whole resource is needed. Not sure if this would work.

Right, I guess there's no way to get the rest of the object without bypassing the cache then.

At least when this is deployed as a standalone container image, we'll be able to limit the cache since we'll have our own Manager. A solution for filtering the cache when using this as a Go module would be nice though.